### PR TITLE
Add Linode server status check.

### PR DESCRIPTION
**Pull request by jalessio**
For bounty: https://assembly.com/runbook/bounties/253

I need some guidance on testing this either manually or adding test code for it. The current state of this pull request is UNTESTED and I'm hoping this will spur conversation on how to test.

Notes:
* I used Digital Ocean droplet status as a template/guide.
* I was unable to find **any** documentation provided by Linode for the meaning of the "STATUS" field in the json response of the [`linode.list` endpoint](https://www.linode.com/api/linode/linode.list). I found two references that more-or-less agree on what the statuses are and I ended up using the text descriptions from the Python library :
  * [Linode Python Bindings](https://github.com/tjfontaine/linode-python/blob/9ee60e8ba0557b68613cf52d8e4a6f926bdc797f/linode/api.py#L386)
  * [Linode Ruby API](https://github.com/incandescent/linode-utils/blob/163842db0c851f5183f80a400215b4012e09d753/lib/linode-utils.rb#L17)
* I could use some guidance on how you'd like to handle error conditions in the monitor. For example, there is definitely the chance of an exception being thrown [in the monitor code](https://github.com/jalessio/cloudroutes-service/compare/asm-products:develop...add_linode_status_monitor?expand=1#diff-630e930faa54fcf17b54f0a3073b8377R44) if the Linode API ever misbehaves, changes, etc. It seems that the [Digital Ocean monitor is subject to the same sort of thing](https://github.com/asm-products/cloudroutes-service/blob/cce850f49b3c3c89a5c96978b181912e603c05b5/src/monitors/checks/digitalocean-status/__init__.py#L42) though so maybe I'm missing something? (perhaps the exceptions are being caught elsewhere in the stack?)

### CODE REVIEW

File: src/monitors/checks/linode-status/__init__.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    if req.status_code < 200 or req.status_code >= 300:
+        line = 'linode-status: Reqeust to {0} sent for monitor {1} - ' \
+               'non-success HTTP code'.format(url, jdata['cid'])
+        logger.warning(line)
+        return False
+
+    # TODO: Add exception handling in case api returns a malformed/unexpected
+    #       json object that does not contain ['DATA'][0]['STATUS'] key.
+    status = req.jsonstatus['DATA'][0]['STATUS']
+    if status in jdata['DATA']['STATUS']:
```
Comment (madflojo) [inline src/monitors/checks/linode-status/__init__.py]: The `jdata['DATA']['STATUS']` should actually be lowercase.

File: src/monitors/checks/linode-status/__init__.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+http_timeout = 5.0
+url = 'https://api.linode.com/'
+
+def check(**kwargs):
+    ''' Perform a Linode info retrieval and check status '''
+    jdata = kwargs['jdata']
+    logger = kwargs['logger']
+    params = {
+        "api_action": "linode.list",
+        "LinodeID": int(redata['data']['linode_id']),
```
Comment (madflojo) [inline src/monitors/checks/linode-status/__init__.py]: In here `linode_id` has an underscore but in the wtforms it does not.

**Comment by madflojo** [comment]
Hey @jalessio,

I added a some notes on a few things that stood out to me. 

As far as your questions go:

* Linode status replies not well documented

Linode's API documentation really doesn't cover this very well. I think where you found the status codes in the python and ruby modules was a great find.

* Error handling

There is [error handling](https://github.com/asm-products/cloudroutes-service/blob/master/src/actions/actioner.py#L116-L124) with where this module is called. So as you have it now if the JSON message was malformed the main process wouldn't crash and burn. With that said though it is of course best to handle any potential errors you may foresee. One thing to note is that with Linode when they have an error they toss it into the `ERRORARRAY` key in their JSON reply but still give a 200 response code.

DO actually gives a non-200 response code if there are errors. If you look at the exception handling you have with the `req.get` statement that is pretty much the model of how errors should be handled.

* Testing

Ah yes testing, this one is tricky. I'm going to break this into two replies, automated unit testing and testing against Linodes actual API.

As far as testing goes @mjhea0 is working on some stuff to get us started with testing monitors and reactions. I think the underlying idea is to use mock to fake a reply from the API. While this works for some basic testing there are obvious problems with this as we are never actually querying their API at that point.

For testing against Linodes actual API, in general the way we have been doing it is to create a free account at the endpoints service and test via the live API. For Linode they don't have a free account so when it gets to the point of testing the actual API I can do that for you and give you any feedback/errors.

I know thats not the best answer but Linode doesn't have any sandbox API and I just don't see any other way of testing against their live API. 

I think there are possibilities with [Mockable.io](http://www.mockable.io/) in the future but from what I can see they kill your end point after a few days. So if we made a fake Linode endpoint and used mock to change the url, we would have to recreate and change the testing url repeatedly. 

For now though you could probably take the example response in Linodes documentation and make a mockable end point for your initial testing at least.

**Comment by jalessio** [comment]
@madflojo For testing, I have this super-basic question:

_what's the best way to manually run a monitor when using the [INSTALLWithDocker.md](https://github.com/asm-products/cloudroutes-service/blob/master/INSTALLWithDocker.md) approach for a development environment?_

I suggest we add your answer to [creating_monitors.md](https://github.com/asm-products/cloudroutes-service/blob/master/docs/developers/creating_monitors.md).

Also, I created a Linode account for testing so that's not an immediate issue although longer term mocking the API responses will probably be necessary for unit tests.

**Comment by mjhea0** [comment]
@jalessio @madflogo I think in the longer term manually testing is the only way to really test this. Adding mocks/fixtures is just too much work and it doesn't really test this all that well. My thought - an army of MTurk workers. Once per month. :)

**Comment by madflojo** [comment]
@jalessio, The simple answer to your question is.

> While mock testing and automated testing is important for future functionality, all monitors or reactions should also be tested manually. This means on your local instance of Runbook you can access the web application and click "Sign Up". Once you sign up for a user you will be logged in and you can create a local test monitor and reaction recipe.

In the future you will probably need to use a user creation script vs. signing up but at the end of the day the idea is the same. The only way to really test this right now is to launch the app and create the monitor. 

@mjhea0, maybe after our weekly release cut off we make a bounty to run QA tests against our pre-production branch? Something like cutoff is on Sunday, Tests are performed by Tuesday, Released into production by Thursday. With that it might make more sense to go to a 2 week release cycle.

**Comment by madflojo** [comment]
Hey @jalessio let me know if you need anymore input on this.

**Pull request merged**